lib/Extension/LanguageServerPhpCodeSniffer/Provider/PhpCodeSnifferDiagnosticsProvider.php

```diff
@@ -0,0 +1,209 @@
+     */
+    private function hasFixableDiagnostics(TextDocumentItem $textDocument): Promise
+    {
+        return \Amp\call(function () use ($textDocument) {
+            $outputJson = yield $this->phpCodeSniffer->diagnose($textDocument, [ '-m' ]);
+
+            $output = json_decode($outputJson, flags: JSON_THROW_ON_ERROR);
+
+            return $output->totals->fixable > 0;
+        });
+    }
+
+    /**
+     * @return Promise<Diagnostic[]|false> False when there are no diagnostics available for file, array othwerwise
+     *                                     Array containing diagnostics to show
+     */
+    private function findDiagnostics(TextDocumentItem $textDocument, CancellationToken $cancel): Promise
+    {
+        return \Amp\call(function () use ($textDocument) {
+            $outputJson = yield $this->phpCodeSniffer->diagnose($textDocument);
+
+            $output = json_decode($outputJson, flags: JSON_THROW_ON_ERROR);
+
+            if (empty($output->files)) {
+                return false;
+            }
+
+            // phpcs return array indexed by file name,
+            // but we only deal with one file, thus don't care about
+            // actual key
+            $files = new ArrayIterator($output->files);
+            $rules = $files->current()->messages;
+
+            $diagnostics = [];
+
+            $diffParser = new Parser();
+
+            foreach ($rules as $rule) {
+                // We treat non-fixable rules as 1 char range.
+                if ($rule->fixable === false) {
+                    $range = new Range(
+                        new Position(--$rule->line, $rule->column),
+                        new Position(--$rule->line, ++$rule->column)
+                    );
+                    $diagnostics[] = yield $this->createRuleDiagnostics($rule, $range);
+                    continue;
+                }
+
+                $sniffWithoutSuffix = $this->getSniffGroup($rule->source);
+                if ($sniffWithoutSuffix === null) {
+                    continue;
+                }
+
+                $fileDiffText = yield $this->phpCodeSniffer->produceFixesDiff($textDocument, [$sniffWithoutSuffix]);
+                $fileDiff = $diffParser->parse($fileDiffText);
+
+                // one file input is passed and one file expected
+                if (count($fileDiff) !== 1) {
+                    $this->logger->warning(
+                        sprintf("Expected phpcs to provide 1 diff, got %s. Skipping diagnostics for file '%s'", count($fileDiff), $textDocument->uri)
+                    );
+
+                    continue;
+                }
+
+                $ranges = $this->rangeForDiff->createRangesForDiff($fileDiff[0]);
+
+                foreach ($ranges as $range) {
+                    $diagnostics[] = yield $this->createRuleDiagnostics($rule, $range);
+                }
+            }
+
+            return $diagnostics;
+        });
+    }
+
+    /**
+     * @return Promise<Diagnostic>
```

Comment:
Phpstan complains about this return, claiming that `\Amp\call` may return `Promise<mixed>` instead of declared type, but the same code for creating rule diagnostic is in php-cs-fixer extension and is not either silenced from phpstan or reported by it.

Any suggestions to help with my confusion would be appeciated.

Reply:
probably ignored in the baseline. Does adding a return type to the anonymous function help? `use (...): Diagnostic`

Reply:
I blindfolded copied php-cs-fixer, which was `yield`ing rules description. This `yield` made a difference. As we don't need anything async in phpcs, I just return object directly and phpstan is happy again (keeping code works as intended).